In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV, train_test_split

In [2]:
df = pd.read_csv("/Users/rifah/Downloads/data.csv")

df.info()
df.shape
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1389 entries, 0 to 1388
Data columns (total 13 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   modified Zurich class            1389 non-null   object
 1   largest spot size                1389 non-null   object
 2   spot distribution                1389 non-null   object
 3   activity                         1389 non-null   int64 
 4   evolution                        1389 non-null   int64 
 5   previous 24 hour flare activity  1389 non-null   int64 
 6   historically-complex             1389 non-null   int64 
 7   became complex on this pass      1389 non-null   int64 
 8   area                             1389 non-null   int64 
 9   area of largest spot             1389 non-null   int64 
 10  common flares                    1389 non-null   int64 
 11  moderate flares                  1389 non-null   int64 
 12  severe flares                    1

activity    evolution  previous 24 hour flare activity  \
count  1389.000000  1389.000000                      1389.000000   
mean      1.150468     2.421166                         1.089993   
std       0.357658     0.617129                         0.403292   
min       1.000000     1.000000                         1.000000   
25%       1.000000     2.000000                         1.000000   
50%       1.000000     2.000000                         1.000000   
75%       1.000000     3.000000                         1.000000   
max       2.000000     3.000000                         3.000000   

       historically-complex  became complex on this pass         area  \
count           1389.000000                  1389.000000  1389.000000   
mean               1.395968                     1.892009     1.025918   
std                0.489234                     0.310481     0.158948   
min                1.000000                     1.000000     1.000000   
25%                1.000000                     2.000000     1.000000   
50%                1.000000                     2.000000     1.000000   
75%                2.000000                     2.000000     1.000000   
max                2.000000                     2.000000     2.000000   

       area of largest spot  common flares  moderate flares  severe flares  
count           1389.000000    1389.000000      1389.000000    1389.000000  
mean               1.175666       0.261339         0.067675       0.009359  
std                0.380673       0.760201         0.353695       0.103534  
min                1.000000       0.000000         0.000000       0.000000  
25%                1.000000       0.000000         0.000000       0.000000  
50%                1.000000       0.000000         0.000000       0.000000  
75%                1.000000       0.000000         0.000000       0.000000  
max                2.000000       8.000000         5.000000       2.000000

In [3]:
df.dtypes

modified Zurich class              object
largest spot size                  object
spot distribution                  object
activity                            int64
evolution                           int64
previous 24 hour flare activity     int64
historically-complex                int64
became complex on this pass         int64
area                                int64
area of largest spot                int64
common flares                       int64
moderate flares                     int64
severe flares                       int64
dtype: object

In [4]:
df.head()

modified Zurich class largest spot size spot distribution  activity  \
0                     C                 S                 O         1   
1                     D                 S                 O         1   
2                     C                 S                 O         1   
3                     D                 S                 O         1   
4                     D                 A                 O         1   

   evolution  previous 24 hour flare activity  historically-complex  \
0          2                                1                     1   
1          3                                1                     1   
2          3                                1                     1   
3          3                                1                     1   
4          3                                1                     1   

   became complex on this pass  area  area of largest spot  common flares  \
0                            2     1                     2              0   
1                            2     1                     2              0   
2                            2     1                     1              0   
3                            2     1                     2              0   
4                            2     1                     2              0   

   moderate flares  severe flares  
0                0              0  
1                0              0  
2                0              0  
3                0              0  
4                0              0

In [5]:

df = df.drop_duplicates()
pd.set_option('display.max_columns', None)
df

modified Zurich class largest spot size spot distribution  activity  \
0                        C                 S                 O         1   
1                        D                 S                 O         1   
2                        C                 S                 O         1   
4                        D                 A                 O         1   
5                        D                 A                 O         1   
...                    ...               ...               ...       ...   
1355                     D                 H                 O         1   
1357                     D                 A                 I         2   
1369                     E                 A                 O         1   
1370                     H                 S                 X         1   
1381                     H                 S                 X         1   

      evolution  previous 24 hour flare activity  historically-complex  \
0             2                                1                     1   
1             3                                1                     1   
2             3                                1                     1   
4             3                                1                     1   
5             2                                1                     1   
...         ...                              ...                   ...   
1355          2                                1                     2   
1357          3                                2                     1   
1369          2                                1                     2   
1370          3                                1                     1   
1381          2                                2                     2   

      became complex on this pass  area  area of largest spot  common flares  \
0                               2     1                     2              0   
1                               2     1                     2              0   
2                               2     1                     1              0   
4                               2     1                     2              0   
5                               2     1                     2              0   
...                           ...   ...                   ...            ...   
1355                            2     1                     1              0   
1357                            2     1                     1              0   
1369                            2     1                     1              4   
1370                            1     1                     1              1   
1381                            2     1                     1              0   

      moderate flares  severe flares  
0                   0              0  
1                   0              0  
2                   0              0  
4                   0              0  
5                   0              0  
...               ...            ...  
1355                0              0  
1357                0              0  
1369                1              0  
1370                0              0  
1381                0              0  

[527 rows x 13 columns]

In [6]:
df.nunique()

modified Zurich class              6
largest spot size                  6
spot distribution                  4
activity                           2
evolution                          3
previous 24 hour flare activity    3
historically-complex               2
became complex on this pass        2
area                               2
area of largest spot               2
common flares                      8
moderate flares                    6
severe flares                      3
dtype: int64

In [7]:
cols = (["modified Zurich class",'largest spot size', 'spot distribution', 'activity', 'evolution',	'previous 24 hour flare activity', 'historically-complex', 'became complex on this pass' ,'area','area of largest spot','common flares', 'moderate flares', 'severe flares'])
for col in cols:
    
    print(df[col].value_counts() )

modified Zurich class
D    164
C    105
E     91
H     87
F     46
B     34
Name: count, dtype: int64
largest spot size
S    156
A    153
K     83
R     73
X     32
H     30
Name: count, dtype: int64
spot distribution
O    196
I    194
X     87
C     50
Name: count, dtype: int64
activity
1    356
2    171
Name: count, dtype: int64
evolution
3    256
2    226
1     45
Name: count, dtype: int64
previous 24 hour flare activity
1    463
3     52
2     12
Name: count, dtype: int64
historically-complex
2    309
1    218
Name: count, dtype: int64
became complex on this pass
2    498
1     29
Name: count, dtype: int64
area
1    493
2     34
Name: count, dtype: int64
area of largest spot
1    373
2    154
Name: count, dtype: int64
common flares
0    347
1    105
2     39
3     19
4      9
5      4
6      3
8      1
Name: count, dtype: int64
moderate flares
0    461
1     51
2      9
4      3
3      2
5      1
Name: count, dtype: int64
severe flares
0    515
1     11
2      1
Name: count, dtype:

In [ ]:
df=pd.get_dummies(df, columns=["modified Zurich class",'largest spot size', 'spot distribution'])

: 

In [ ]:
X = df.drop(['common flares', 'moderate flares', 'severe flares'], axis=1)
y = df[['common flares', 'moderate flares', 'severe flares']]

ran = RandomForestRegressor()

grid_params1 = {"n_estimators": [ 50 ,75, 100, 150, 175], 
               "criterion": ["squared_error", "absolute_error", "friedman_mse"],
               "max_depth" : [10,15,20,None],
               "min_samples_split": [6,10,14],
               "min_samples_leaf": [1,3,5],
               "n_jobs": [-1],
                }
rancv = GridSearchCV(ran, grid_params1, scoring="neg_mean_squared_error", cv=4, refit=True, verbose=7, return_train_score=True, error_score='raise')

X_train, X_test,  y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

rancv.fit(X_train, y_train)



In [ ]:
print(rancv.best_score_)

In [ ]:
from sklearn.linear_model import RidgeCV 

grid_params2 = {
    "alpha": [0,0.1, 0.5, 0.7, 1, 1.5, 1.7, 2, 5,6,7,8,9,10,
              ]
}

ridgecvval = RidgeCV(alphas=grid_params2["alpha"], fit_intercept=True, scoring="neg_mean_squared_error", cv=4, alpha_per_target=False)
ridgecvval.fit(X_train,y_train)
print(ridgecvval.best_score_)


In [ ]:
from sklearn.neighbors import KNeighborsRegressor

knn = KNeighborsRegressor()

grid_params3 = {
    "n_neighbors" : [2,5,7,8,10,12,15],
    "weights" : ["uniform", "distance", None],
    "algorithm": ['auto', 'ball_tree', 'kd_tree', 'brute'],
    "leaf_size" : [10,30,50,75],
    "p": [1, 1.5, 2,3],
    "n_jobs": [-1]
}

knncv = GridSearchCV(knn, grid_params3, scoring="neg_mean_squared_error", cv=4, refit=True, verbose=7, return_train_score=True, error_score='raise')
knncv.fit(X_train, y_train)
print(knncv.best_score_)

In [ ]:
print("Negative mean squared errors in training data ")
print(f"Random Forest Regressor {rancv.best_score_}")
print(f"Ridge Regression {ridgecvval.best_score_}")
print(f"K Nearest Regressor {knncv.best_score_}")


In [ ]:
print("Testing each model on testing data")

bestran = rancv.best_estimator_
bestknn = knncv.best_estimator_

y1pred = bestran.predict(X_test)
y2pred = ridgecvval.predict(X_test)
y3pred = bestknn.predict(X_test)

from sklearn.metrics import mean_squared_error, mean_absolute_error

print("MSE Scores after training each model")
print("Random forest ",mean_squared_error(y_test,y1pred))
print("Ridge regression ",mean_squared_error(y_test,y2pred))
print("K nearest neighbour regressor ",mean_squared_error(y_test,y3pred))

print("MAE Scores after training each model")
print("Random forest ",mean_absolute_error(y_test,y1pred))
print("Ridge regression ",mean_absolute_error(y_test,y2pred))
print("K nearest neighbour regressor ",mean_absolute_error(y_test,y3pred))



Testing each model on testing data
MSE Scores after training each model
Random forest  0.5224302442599035
Ridge regression  0.5182689299222764
K nearest neighbour regressor  0.5507617051013278
MAE Scores after training each model
Random forest  0.3392877716119463
Ridge regression  0.3376033576376883
K nearest neighbour regressor  0.35408805031446544


Data - previous cell

Testing each model on testing data
MSE Scores after training each model
Random forest  0.5224302442599035
Ridge regression  0.5182689299222764
K nearest neighbour regressor  0.5507617051013278
MAE Scores after training each model
Random forest  0.3392877716119463
Ridge regression  0.3376033576376883
K nearest neighbour regressor  0.35408805031446544


Hence Ridge Regression will be the model deployed